rsocket-transport-netty/src/main/java/io/rsocket/transport/netty/server/WebsocketRouteTransport.java

```diff
@@ -88,4 +101,121 @@ public WebsocketRouteTransport(
+    private final HashMap<String, Matcher> matchers = new HashMap<>();
+    private final HashMap<String, Map<String, String>> vars = new HashMap<>();
+
+    private final Pattern uriPattern;
+
+    static String filterQueryParams(String uri) {
+      int hasQuery = uri.lastIndexOf("?");
+      if (hasQuery != -1) {
+        return uri.substring(0, hasQuery);
+      } else {
+        return uri;
+      }
+    }
+
+    /**
+     * Creates a new {@code UriPathTemplate} from the given {@code uriPattern}.
+     *
+     * @param uriPattern The pattern to be used by the template
+     */
+    UriPathTemplate(String uriPattern) {
+      String s = "^" + filterQueryParams(uriPattern);
+
+      Matcher m = NAME_SPLAT_PATTERN.matcher(s);
+      while (m.find()) {
+        for (int i = 1; i <= m.groupCount(); i++) {
+          String name = m.group(i);
+          pathVariables.add(name);
+          s = m.replaceFirst(NAME_SPLAT_REPLACEMENT.replaceAll("%NAME%", name));
+          m.reset(s);
+        }
+      }
+
+      m = NAME_PATTERN.matcher(s);
+      while (m.find()) {
+        for (int i = 1; i <= m.groupCount(); i++) {
+          String name = m.group(i);
+          pathVariables.add(name);
+          s = m.replaceFirst(NAME_REPLACEMENT.replaceAll("%NAME%", name));
+          m.reset(s);
+        }
+      }
+
+      m = FULL_SPLAT_PATTERN.matcher(s);
+      while (m.find()) {
+        s = m.replaceAll(FULL_SPLAT_REPLACEMENT);
+        m.reset(s);
+      }
+
+      this.uriPattern = Pattern.compile(s + "$");
+    }
+
+    /**
+     * Tests the given {@code uri} against this template, returning {@code true} if the uri matches
+     * the template, {@code false} otherwise.
+     *
+     * @param uri The uri to match
+     * @return {@code true} if there's a match, {@code false} otherwise
+     */
+    public boolean matches(String uri) {
+      return matcher(uri).matches();
+    }
+
+    /**
+     * Matches the template against the given {@code uri} returning a map of path parameters
+     * extracted from the uri, keyed by the names in the template. If the uri does not match, or
+     * there are no path parameters, an empty map is returned.
+     *
+     * @param uri The uri to match
+     * @return the path parameters from the uri. Never {@code null}.
+     */
+    final Map<String, String> match(String uri) {
+      Map<String, String> pathParameters = vars.get(uri);
+      if (null != pathParameters) {
+        return pathParameters;
+      }
+
+      pathParameters = new HashMap<>();
+      Matcher m = matcher(uri);
+      if (m.matches()) {
+        int i = 1;
+        for (String name : pathVariables) {
+          String val = m.group(i++);
+          pathParameters.put(name, val);
+        }
+      }
+      synchronized (vars) {
+        vars.put(uri, pathParameters);
+      }
+
+      return pathParameters;
+    }
+
+    private Matcher matcher(String uri) {
+      uri = filterQueryParams(uri);
+      Matcher m = matchers.get(uri);
```

Comment:
feels like this should be concurrenthashmap.computeIfAbsent

Reply:
fixed